internal/api/v1beta1/permission_check.go

```diff
@@ -22,6 +22,76 @@ func (h Handler) CheckResourcePermission(ctx context.Context, req *shieldv1beta1
 	//	logger.Error(formattedErr.Error())
 	//	return nil, status.Errorf(codes.NotFound, formattedErr.Error())
 	//}
+	//  To have backward compatibility
+	if req.ObjectId != "" && len(req.ResourcePermissions) == 0 {
+		return h.checkSingleResourcePermission(ctx, req)
+	}
+
+	if len(req.ResourcePermissions) > h.maxBulkCheckLimit || len(req.ResourcePermissions) == 0 {
+		formattedErr := fmt.Errorf("%s: %s", ErrRequestBodyValidation, "resource_permissions")
+		logger.Error(formattedErr.Error())
+		return nil, formattedErr
+	}
+
+	var results []*shieldv1beta1.CheckResourcePermissionResponse_ResourcePermissionResponse
+	bulkCheckCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	resultCh := make(chan *shieldv1beta1.CheckResourcePermissionResponse_ResourcePermissionResponse)
+	defer close(resultCh)
+	errorCh := make(chan error)
+	defer close(errorCh)
+
+	for _, resourcePermission := range req.ResourcePermissions {
+		go func(bulkCheckCtx context.Context, resourcePermission *shieldv1beta1.ResourcePermission) {
+			result, err := h.resourceService.CheckAuthz(ctx, resource.Resource{
+				Name:        resourcePermission.GetObjectId(),
+				NamespaceID: resourcePermission.GetObjectNamespace(),
+			}, action.Action{ID: resourcePermission.GetPermission()})
+			if err != nil {
+				switch {
+				case errors.Is(err, user.ErrInvalidEmail):
+					errorCh <- grpcUnauthenticated
+					return
+				default:
+					formattedErr := fmt.Errorf("%s: %w", ErrInternalServer, err)
+					logger.Error(formattedErr.Error())
+					errorCh <- status.Errorf(codes.Internal, ErrInternalServer.Error())
+					return
+				}
+			}
+			select {
+			case <-bulkCheckCtx.Done():
+				return
+			case resultCh <- &shieldv1beta1.CheckResourcePermissionResponse_ResourcePermissionResponse{
+				ObjectId:        resourcePermission.GetObjectId(),
+				ObjectNamespace: resourcePermission.GetObjectNamespace(),
+				Permission:      resourcePermission.GetPermission(),
+				Allowed:         result,
+			}:
+			}
+		}(bulkCheckCtx, resourcePermission)
```

Comment:
why resultCh not passed?

Reply:
decided to use it, because those values won't be modified like `resourcePermission`. However, modified the func to take channel arguments as well, as to keep the code consistent.